ow_power_system/src/power_system_node.cpp

```diff
@@ -13,92 +13,116 @@
+    float max_voltage = (baseline_voltage + voltage_range) + (battery_lifetime - update_count) / battery_lifetime * 0.8; // [V]
+    
+    // If voltage limits dip below baseline, set to baseline values
+    if (min_voltage < baseline_voltage)
     {
-      ROS_ERROR_NAMED("power_system_node", "power_update_rate param was set to zero! passed value ignored.");
+      min_voltage = baseline_voltage;
     }
-    else
+    if (max_voltage < (baseline_voltage + voltage_range))
     {
-      power_update_rate = power_update_rate_override;
+      max_voltage = baseline_voltage + voltage_range;
     }
-  }
-
-  ROS_INFO_STREAM_NAMED("power_system_node", "power_update_rate is set to: " << power_update_rate << " Hz");
-  
-  // ROS Loop. Note that once this loop starts,
-  // this function (and node) is terminated with an interrupt.
-  
-  ros::Rate rate(power_update_rate);
-  //individual soc_msg to be published by SOC_pub
-  std_msgs::Int16 rul_msg;
-  rul_msg.data = 0;  // immediately set to a good default
-  std_msgs::Float64 soc_msg;
-  soc_msg.data = 0.0;
-  std_msgs::Float64 tempbat_msg;
-  tempbat_msg.data = 0.0;
-  int temp_index = 1; // Set to 1 for now (constant vector). This will change to median SOC or RUL index or fixed percentile
-  ROS_INFO ("Power system node running");
-
-  while (ros::ok()) {
 
+    // Voltage estimate based on pseudorandom noise and moving range
+    float voltage_est = min_voltage + static_cast <float> (rand()) / (static_cast <float> (RAND_MAX/(max_voltage - min_voltage)));
+    
+    // Create single line csv containing power data and voltage/temp estimates
+    string csv_path;
+    string default_csv = "/data/gsap_input_data.csv";
+    bool csv_path_param_exist = nh.param("power_draw_csv_path", csv_path,
+      ros::package::getPath("ow_power_system") + default_csv);
+
+    // Open new file
+    std::ofstream myfile;
+    myfile.open (csv_path);
+    myfile << "Timestamp, power, temperature, voltage\n";
+    // First line is used to initialize GSAP at first time step
+    myfile << update_count << ", " << init_power << ", " << init_temp << ", " << init_voltage << "\n";
+    // Update time step and GSAP initialization parameters
+    update_count = update_count + 1.0;
+    init_power = mech_power;
+    init_temp = temp_est;
+    init_voltage = voltage_est;
+    // Line used to make the prediction (use current time step)
+    myfile << update_count << ", " << mech_power << ", " << temp_est << ", " << voltage_est << "\n";
+    myfile.close();
+
+    //individual msgs to be published
+    std_msgs::Int16 rul_msg;
+    rul_msg.data = 0;  // immediately set to a good default
+    std_msgs::Float64 soc_msg;
+    soc_msg.data = 0.0;
+    std_msgs::Float64 tempbat_msg;
+    tempbat_msg.data = 0.0;
+
+    int temp_index = 1; // Set to 1 for now (constant vector). This will change to median SOC or RUL index or fixed percentile
+
+    // Read battery data from newly created file.
+    auto power_data = read_file(csv_path);
```

Comment:
I know exactly what you doing here and although it makes good re-use of existing code this is considered lazy programming :facepalm: